packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+  private queue = new PriorityQueue({concurrency: 1});
+
+  // Memory tracking wallet nonce to assist with retries
+  private memoryNonce = 0;
+
+  // TODO: Remove reference to this object, and instead replace with
+  // outbox based communication. Right now used to call correct callbacks
+  private channelWallet: ChannelWallet | undefined = undefined;
+
+  constructor(env: OnchainServiceEnvironment) {
+    this.retries = env.transactionRetries || DEFAULT_MAX_RETRIES;
+    this.provider =
+      typeof env.provider === 'string' ? new providers.JsonRpcProvider(env.provider) : env.provider;
+    this.wallet =
+      typeof env.wallet === 'string'
+        ? new Wallet(env.wallet, this.provider)
+        : (env.wallet as Wallet).connect(this.provider);
+  }
+
+  public attachChannelWallet(wallet: ChannelWallet): void {
+    this.channelWallet = wallet;
+  }
+
+  /**
+   * Adds a channel to watch events and submit transactions for
+   *
+   * @param channelId Unique channel identifier to register chain listeners for
+   * @param assetHolders Onchain addresses of the channel chain context
+   * @returns String if there is an error, undefined if successful
+   */
+  public registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderInformation[]
+  ): Promise<string | undefined> {
+    if (!this.channelWallet) {
+      throw new Error(
+        'Must attach channel wallet before registering channel, this will go away in v1'
+      );
+    }
+
+    // If the channel has already been registered, return
+    if (this.store.has(channelId)) {
+      return Promise.resolve(undefined);
+    }
+
+    // Create and store new contracts if we don't have record of
+    // required assetHolders
+    assetHolders.forEach(assetHolder => {
+      if (this.assetHolders.has(assetHolder.contractAddress)) {
+        return;
+      }
+      // TODO: Use the AssetHolder abi?
+      const contract = new Contract(
+        assetHolder.contractAddress,
+        ContractArtifacts.EthAssetHolderArtifact.abi,
+        this.provider
+      );
+      this._registerAssetHolderCallbacks(contract);
+      this.assetHolders.set(assetHolder.contractAddress, {...assetHolder, contract});
+    });
+
+    // Add the channel to service storage
+    this.store.set(channelId, {assetHolders, events: {}});
+
+    return Promise.resolve(undefined);
+  }
+
+  public async submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse> {
+    if (!this.channelWallet) {
+      throw new Error(
+        'Must attach channel wallet before registering channel, this will go away in v1'
+      );
+    }
+
+    // If the channel is not registered, do not send transactions
+    if (!this.store.has(channelId)) {
+      throw new Error(`Channel is not registered ${channelId}`);
+    }
+
+    // Create the errors object
+    const errors: {[k: number]: string} = {};
+    for (let attempt = 1; attempt < this.retries; attempt += 1) {
```

Comment:
Might be a touch cleaner
```suggestion
    for (const attempt in _.range(retries - 1)) {
```